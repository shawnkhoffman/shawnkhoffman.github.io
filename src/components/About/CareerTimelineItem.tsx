interface CareerTimelineItemProps {
  date: string;
  title: string;
  description: string;
  isEven?: boolean;
}

export default function CareerTimelineItem({ date, title, description, isEven = false }: CareerTimelineItemProps) {
  return (
    <li data-testid="timeline-item" role="listitem" tabIndex={0}>
      <div className="timeline-middle">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 20 20"
          fill="currentColor"
          className="h-5 w-5"
          data-testid="decorative-icon"
          aria-hidden="true"
        >
          <path
            fillRule="evenodd"
            d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
            clipRule="evenodd"
          />
        </svg>
      </div>
      <div className={`${isEven ? 'timeline-end' : 'timeline-start'} mb-10 md:${isEven ? 'text-start' : 'text-end'}`}>
        <time className="font-mono italic">{date}</time>
        <h3 className="text-lg font-black">{title}</h3>
        <p data-testid="timeline-description">{description}</p>
      </div>
      <hr />
    </li>
  );
}
