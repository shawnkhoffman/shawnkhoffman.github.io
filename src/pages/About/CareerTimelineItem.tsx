import React from 'react';

export interface CareerTimelineItemProps {
  date: string;
  title: string;
  description: string;
  position: 'start' | 'end';
  testId?: string;
}

const CareerTimelineItem: React.FC<CareerTimelineItemProps> = ({ 
  date, 
  title, 
  description, 
  position,
  testId = 'timeline-item'
}) => {
  return (
    <li 
      data-testid={testId}
      tabIndex={0}
      role="listitem"
    >
      <div className="timeline-middle">
        <svg
          data-testid="decorative-icon"
          aria-hidden="true"
          className="h-5 w-5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            fillRule="evenodd"
            d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
            clipRule="evenodd"
          />
        </svg>
      </div>
      <div className={`timeline-${position} mb-10 md:text-${position === 'start' ? 'end' : 'start'}`}>
        <time className="font-mono italic">{date}</time>
        <h3 className="text-lg font-black">{title}</h3>
        <p role="article">{description}</p>
      </div>
      <hr />
    </li>
  );
};

export default CareerTimelineItem;
