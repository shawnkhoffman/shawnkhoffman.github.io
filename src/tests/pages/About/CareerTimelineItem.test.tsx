import { cleanup, render, screen, within } from '@testing-library/react';
import { describe, it, expect, beforeEach } from 'vitest';
import CareerTimelineItem from '@/pages/About/CareerTimelineItem';

describe('CareerTimelineItem', () => {
  const createTestId = (testName: string) => `timeline-item-${testName}`;
  
  beforeEach(() => {
    cleanup();
  });

  describe('Rendering', () => {
    it('renders without crashing', () => {
      render(<CareerTimelineItem date="2024" title="Test Title" description="Test Description" position="start" testId={createTestId('rendering')} />);
      expect(screen.getByTestId(createTestId('rendering'))).toBeInTheDocument();
    });

    it('matches snapshot', () => {
      const { container } = render(<CareerTimelineItem date="2024" title="Test Title" description="Test Description" position="start" testId={createTestId('snapshot')} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Structure', () => {
    it('contains all required elements', () => {
      const testId = createTestId('structure');
      render(<CareerTimelineItem 
        date="2024"
        title="Test Title"
        description="Test Description"
        position="start"
        testId={testId}
      />);
      
      const timelineItem = screen.getByTestId(testId);
      const timeElement = within(timelineItem).getByText('2024');
      const titleElement = within(timelineItem).getByText('Test Title');
      const descriptionElement = within(timelineItem).getByText('Test Description');
      
      expect(timeElement).toBeInTheDocument();
      expect(titleElement).toBeInTheDocument();
      expect(descriptionElement).toBeInTheDocument();
    });
  });

  describe('Interactions', () => {
    it('is keyboard navigable', () => {
      const testId = createTestId('keyboard');
      render(<CareerTimelineItem 
        date="2024"
        title="Test Title"
        description="Test Description"
        position="start"
        testId={testId}
      />);
      
      const timelineItem = screen.getByTestId(testId);
      
      expect(timelineItem).toHaveAttribute('tabIndex', '0');
      expect(timelineItem).toHaveAttribute('role', 'listitem');
      
      timelineItem.focus();
      expect(timelineItem).toHaveFocus();
    });

    it('remains visible at different viewport sizes', () => {
      const testId = createTestId('viewport');
      render(<CareerTimelineItem 
        date="2024"
        title="Test Title"
        description="Test Description"
        position="start"
        testId={testId}
      />);
      
      const timelineItem = screen.getByTestId(testId);
      
      window.innerWidth = 375;
      window.dispatchEvent(new Event('resize'));
      expect(timelineItem).toBeVisible();
      
      window.innerWidth = 1024;
      window.dispatchEvent(new Event('resize'));
      expect(timelineItem).toBeVisible();
    });
  });

  describe('Accessibility', () => {
    it('has decorative elements marked appropriately', () => {
      const testId = createTestId('decorative');
      render(<CareerTimelineItem 
        date="2024"
        title="Test Title"
        description="Test Description"
        position="start"
        testId={testId}
      />);
      
      const timelineItem = screen.getByTestId(testId);
      const svg = within(timelineItem).getByTestId('decorative-icon');
      expect(svg).toHaveAttribute('aria-hidden', 'true');
    });

    it('maintains a logical reading order', () => {
      const testId = createTestId('reading-order');
      render(<CareerTimelineItem 
        date="2024"
        title="Test Title"
        description="Test Description"
        position="start"
        testId={testId}
      />);
      
      const timelineItem = screen.getByTestId(testId);
      const content = timelineItem.textContent;
      expect(content).toMatch(/2024.*Test Title.*Test Description/);
    });
  });
});
