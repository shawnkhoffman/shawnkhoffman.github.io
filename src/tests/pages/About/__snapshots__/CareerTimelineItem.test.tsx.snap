// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CareerTimelineItem > Interactions > remains visible at different viewport sizes 1`] = `
<li
  data-testid="timeline-item-snapshot"
  role="listitem"
  tabindex="0"
>
  <div
    class="timeline-middle"
  >
    <svg
      aria-hidden="true"
      class="h-5 w-5"
      data-testid="decorative-icon"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <div
    class="timeline-start mb-10 md:text-end"
  >
    <time
      class="font-mono italic"
    >
      2024
    </time>
    <h3
      class="text-lg font-black"
    >
      Test Title
    </h3>
    <p
      role="article"
    >
      Test Description
    </p>
  </div>
  <hr />
</li>
`;
